nixos/postsrd: settings option, hardening, postfix integration

mweinelt opened the pull request:
See individual commit messages.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
nixos/postsrsd: migrate to rfc42 settings

Allow a freeform configuration approach to satisfy different
configuration complexities.

Remove confinement options and make its hardening options more explicit
and removed the deprecated PermissionStartOnly= option.

mweinelt committed:
nixos/postsrsd: harden and modernize systemd unit

This replaces the previous confinement settings with a more complete and
context-sensitive hardening setup.

Also exposes the current config at /etc/postsrsd.conf, which makes it
easily inspectable.

mweinelt committed:
nixos/postsrsd: integrate with postfix by default

SuperSandro2000 commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -62,24 +87,117 @@ in
+              example = "inet:localhost:10003";
+              description = ''
+                Path to the Unix socket for connecting to postsrsd.
+                Read-only, intended for usage when integrating postsrsd into other NixOS config.'';
```
```suggestion
                Read-only, intended for usage when integrating postsrsd into other NixOS config.
              '';
```

SuperSandro2000 commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -143,15 +252,21 @@ in
       before = [ "postfix.service" ];
       wantedBy = [ "multi-user.target" ];
       requires = [ "postsrsd-generate-secrets.service" ];
-      confinement.enable = true;
```
I think confinement is doing more than just those hardening options and limits what files the process can read. It kinda makes debugging harder but it would be fine for me.

mweinelt replied:
Yeah, sets `RootDirectory=` and lots of `BindReadOnlyPaths=`. Not sure if that is worth the trouble.

SuperSandro2000 commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -252,21 +254,56 @@ in
       serviceConfig = {
-        ExecStart = "${lib.getExe pkgs.postsrsd} -C ${configFile}";
+        ExecStart = toString [
+          (lib.getExe pkgs.postsrsd)
```
Can we add a package option?

SuperSandro2000 commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -200,6 +200,14 @@ in
         '';
       };
 
+      configurePostfix = lib.mkOption {
```
Should we assert that postfix is enabled?

mweinelt replied:
The resulting configuration is already guarded around `config.services.postfix.enable`. This is like it is in tlsrpt and postfix-tlspol.

jfly commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 424954`
Commit: `4b0006d3e83fc5012db2ae1467d53ab041297d53`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

jfly reviewed (approved):
Thanks for doing this! Sorry for all the nit-y comments. I'm giving you a shipit because I trust you to resolve/ignore them as you see fit. I don't think I saw anything terribly important, but *full disclosure* I completely ignored the systemd hardening stuff, so I'm just trusting you that this is better than the `confinement.enable = true` we were doing before.

I'll deploy this patch to my personal mailserver and report back.

jfly reviewed (commented):
I just tried deploying this to my personal mailserver. The `srs-domain` setting caused trouble for me (I left a comment inline), but otherwise things built and deployed just fine. Unfortunately, I don't seem to be able to send email. Here's what I see in syslog:

```
Jul 14 17:31:10 doli postfix/smtpd[227119]: warning: connect to /run/postsrsd/socket: Permission denied
Jul 14 17:31:10 doli postfix/smtpd[227119]: warning: table socketmap:unix:/run/postsrsd/socket:reverse lookup error: Permission denied
Jul 14 17:31:10 doli postfix/smtpd[227119]: warning: socketmap:unix:/run/postsrsd/socket:reverse lookup error for "[redacted]"
```

jfly commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -214,97 +222,114 @@ in
-    users.groups = lib.optionalAttrs (cfg.group == "postsrsd") {
-      postsrsd.gid = config.ids.gids.postsrsd;
-    };
+      systemd.services.postfix.serviceConfig.SupplementaryGroups = [ cfg.group ];
```
That makes sense. I can't seem to find the button to mark this resolved.

mweinelt committed:
nixos/postsrsd: add package option, migrate enable option

mweinelt commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -62,24 +87,117 @@ in
+              '';
+            };
+
+            separator = lib.mkOption {
```
This adds a useful eval time type check while advertising the default. I think it is not harmful to keep it around and add the upstream notice about not changing it.

mweinelt commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -62,24 +87,117 @@ in
-        default = "=";
-        description = "First separator character in generated addresses";
+                ::: {.warning}
+                For privacy reasons the domain used for SRS should not be used as a regular mail domain!
```
Added the "multiple unrelated mail domains" aspect.

jfly commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -214,97 +222,114 @@ in
-        uid = config.ids.uids.postsrsd;
+  config = lib.mkMerge [
+    (lib.mkIf (cfg.enable && cfg.configurePostfix && config.services.postfix.enable) {
+      services.postfix.config = lib.mkIf (cfg.configurePostfix && config.services.postfix.enable) {
```
Sorry, I'm not very familiar with `lib.mkMerge`. My brain is not seeing how it affects this duplicity (see screenshot):

<img width="1279" height="234" alt="image" src="https://github.com/user-attachments/assets/012b35dd-9c00-410f-881f-f85594f691a2" />

(not a blocker, just trying to learn)

mweinelt replied:
Oops.

mweinelt replied:
Fixed.

mweinelt commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -62,24 +87,125 @@ in
+              '';
+            };
+
+            srs-domain = lib.mkOption {
```
The srs-domain field was nullable, but I forgot filtering out null values in the resulting config. Fixed.

mweinelt commented:
> I just tried deploying this to my personal mailserver. The `srs-domain` setting caused trouble for me (I left a comment inline), but otherwise things built and deployed just fine. Unfortunately, I don't seem to be able to send email. Here's what I see in syslog:
> 
> ```
> Jul 14 17:31:10 doli postfix/smtpd[227119]: warning: connect to /run/postsrsd/socket: Permission denied
> Jul 14 17:31:10 doli postfix/smtpd[227119]: warning: table socketmap:unix:/run/postsrsd/socket:reverse lookup error: Permission denied
> Jul 14 17:31:10 doli postfix/smtpd[227119]: warning: socketmap:unix:/run/postsrsd/socket:reverse lookup error for "[redacted]"
> ```

Fixed. SupplementaryGroups cannot work on the postfix unit, because the child processes don't inherit the group membership.

Please retest.

jfly commented:
> Please retest.

Thanks. I'm out of time for today, but will do this tomorrow.

SuperSandro2000 commented on nixos/modules/services/mail/postsrsd.nix:
```diff
@@ -93,66 +230,120 @@ in
+        sender_canonical_maps = "socketmap:${cfg.settings.socketmap}:forward";
+        sender_canonical_classes = "envelope_sender";
+        recipient_canonical_maps = "socketmap:${cfg.settings.socketmap}:reverse";
+        recipient_canonical_classes = [
```
This is probably related to the postfix module.

```
       error: The option `services.postfix.config.recipient_canonical_classes' is defined multiple times while it's expected to be unique.

       Definition values:
       - In `/nix/store/qd6rcfggxm1z6lqrj9riz9zd0q500d15-source/hosts/mail': "envelope_recipient, header_recipient"
       - In `/nix/store/p2fs4zm38yji76hk93birlyw97b7ikik-source/nixos/modules/services/mail/postsrsd.nix':
           [
             "envelope_recipient"
             "header_recipient"
           ]
       Use `lib.mkForce value` or `lib.mkDefault value` to change the priority on any of these definitions.
```

SuperSandro2000 replied:
That seems to be really unmaintained. Last commit is 9 years ago. Should we drop it in another PR? https://github.com/Fruneau/pfixtools

mweinelt committed:
nixos/pfix-srsd: migrate postfix integration from postfix module

The postfix module is too big to host every individual integration option
and moving it here has no downside.

SuperSandro2000 reviewed (approved):
otherwise seems to work for me

jfly reviewed (approved):
I haven't re-reviewed the diff, but I did just apply this patch ([diff](https://github.com/jfly/snow/compare/update-postsrsd?expand=1)) to my mailserver and can still send email!
